File: server/src/pipeline/runners.ts
import { z } from "zod";
import path from "node:path";
import { db, nowIso, MEDIA_DIR } from "../db.js";
import { newId, extractJson } from "../util.js";
import { generateText } from "../providers/text.js";
import { generateImage } from "../providers/image.js";
import { generateVideo } from "../providers/video.js";
import * as P from "./prompts.js";

// ---------- zod schemas for LLM output ----------

const DialogueSchema = z.object({ speaker: z.string(), line: z.string() });
const SceneSchema = z.object({
  location: z.string(),
  timeOfDay: z.string().default(""),
  action: z.string(),
  dialogues: z.array(DialogueSchema).default([]),
});
const ScriptOutSchema = z.object({
  episodes: z
    .array(
      z.object({
        title: z.string(),
        synopsis: z.string().default(""),
        scenes: z.array(SceneSchema).min(1),
      }),
    )
    .min(1),
});

const AssetsOutSchema = z.object({
  assets: z
    .array(
      z.object({
        kind: z.enum(["character", "scene"]),
        name: z.string().min(1),
        description: z.string().default(""),
        imagePrompt: z.string().default(""),
      }),
    )
    .min(1),
});

const ShotsOutSchema = z.object({
  shots: z
    .array(
      z.object({
        description: z.string(),
        camera: z.string().default(""),
        dialogue: z.string().default(""),
        assetNames: z.array(z.string()).default([]),
        imagePrompt: z.string().min(1),
        videoPrompt: z.string().default(""),
      }),
    )
    .min(1),
});

/** 调 LLM 并解析结构化输出；解析失败带错误反馈自愈重试一次 */
async function structuredText<T>(system: string, user: string, schema: z.ZodType<T>): Promise<T> {
  const first = await generateText(system, user);
  try {
    return schema.parse(extractJson(first.content));
  } catch (e1) {
    const err1 = e1 instanceof Error ? e1.message : String(e1);
    const second = await generateText(system, P.repairUser(user, first.content, err1.slice(0, 300)));
    try {
      return schema.parse(extractJson(second.content));
    } catch (e2) {
      const err2 = e2 instanceof Error ? e2.message : String(e2);
      throw new Error(
        `模型输出两次都无法解析。最后错误: ${err2.slice(0, 300)}。原始输出开头: ${second.content.slice(0, 200)}`,
      );
    }
  }
}

// ---------- job runners（每个对应一种 job.kind）----------

export type JobRow = {
  id: string;
  projectId: string;
  kind: string;
  targetId: string;
  payload: string;
};

function getProject(projectId: string) {
  const p = db.prepare("SELECT * FROM projects WHERE id=?").get(projectId) as
    | { id: string; name: string; artStyle: string }
    | undefined;
  if (!p) throw new Error("项目不存在（可能已被删除）");
  return p;
}

/** 小说 → 分集剧本。成功后替换该项目全部 episodes（及其 shots）。 */
export async function runScriptGen(job: JobRow): Promise<string> {
  const project = getProject(job.projectId);
  const payload = JSON.parse(job.payload) as { episodeCount?: number };
  const novel = db.prepare("SELECT * FROM novels WHERE projectId=?").get(job.projectId) as
    | { title: string; content: string }
    | undefined;
  if (!novel || !novel.content.trim()) throw new Error("请先导入小说");

  const episodeCount = Math.min(Math.max(payload.episodeCount ?? 3, 1), 12);
  const out = await structuredText(
    P.SCRIPT_GEN_SYSTEM,
    P.scriptGenUser(novel.title || project.name, novel.content, episodeCount),
    ScriptOutSchema,
  );

  const replace = db.transaction(() => {
    db.prepare("DELETE FROM episodes WHERE projectId=?").run(job.projectId);
    const ins = db.prepare(
      "INSERT INTO episodes (id, projectId, idx, title, synopsis, scriptJson, createdAt) VALUES (?,?,?,?,?,?,?)",
    );
    out.episodes.forEach((ep, i) => {
      ins.run(newId(), job.projectId, i + 1, ep.title, ep.synopsis, JSON.stringify(ep.scenes), nowIso());
    });
  });
  replace();
  return `生成 ${out.episodes.length} 集剧本`;
}

/** 全部剧本 → 角色/场景资产列表（不含图片）。按名称合并，已有的保留图片。 */
export async function runAssetExtract(job: JobRow): Promise<string> {
  const project = getProject(job.projectId);
  const episodes = db
    .prepare("SELECT title, synopsis, scriptJson FROM episodes WHERE projectId=? ORDER BY idx")
    .all(job.projectId) as { title: string; synopsis: string; scriptJson: string }[];
  if (episodes.length === 0) throw new Error("请先生成剧本");

  const summary = episodes
    .map((e, i) => `第${i + 1}集《${e.title}》：${e.synopsis}\n${e.scriptJson}`)
    .join("\n\n");
  const out = await structuredText(
    P.ASSET_EXTRACT_SYSTEM,
    P.assetExtractUser(summary, project.artStyle),
    AssetsOutSchema,
  );

  let added = 0;
  let updated = 0;
  const upsert = db.transaction(() => {
    for (const a of out.assets) {
      const existing = db
        .prepare("SELECT id FROM assets WHERE projectId=? AND name=? AND kind=?")
        .get(job.projectId, a.name, a.kind) as { id: string } | undefined;
      if (existing) {
        db.prepare("UPDATE assets SET description=?, imagePrompt=? WHERE id=?").run(
          a.description,
          a.imagePrompt,
          existing.id,
        );
        updated++;
      } else {
        db.prepare(
          "INSERT INTO assets (id, projectId, kind, name, description, imagePrompt, status, createdAt) VALUES (?,?,?,?,?,?,'draft',?)",
        ).run(newId(), job.projectId, a.kind, a.name, a.description, a.imagePrompt, nowIso());
        added++;
      }
    }
  });
  upsert();
  return `提取资产：新增 ${added} 个，更新 ${updated} 个`;
}

/** 单集剧本 → 分镜镜头列表 */
export async function runStoryboardGen(job: JobRow): Promise<string> {
  const project = getProject(job.projectId);
  const episode = db.prepare("SELECT * FROM episodes WHERE id=?").get(job.targetId) as
    | { id: string; title: string; scriptJson: string }
    | undefined;
  if (!episode) throw new Error("剧集不存在");

  const assets = db
    .prepare("SELECT kind, name, description FROM assets WHERE projectId=?")
    .all(job.projectId) as { kind: string; name: string; description: string }[];
  const assetsContext =
    assets.length > 0
      ? assets.map((a) => `[${a.kind === "character" ? "角色" : "场景"}] ${a.name}：${a.description}`).join("\n")
      : "（尚未提取资产，请根据剧本自行保持角色外观一致）";

  const out = await structuredText(
    P.STORYBOARD_GEN_SYSTEM,
    P.storyboardGenUser(episode.title, episode.scriptJson, assetsContext, project.artStyle),
    ShotsOutSchema,
  );

  const replace = db.transaction(() => {
    db.prepare("DELETE FROM shots WHERE episodeId=?").run(episode.id);
    const ins = db.prepare(
      `INSERT INTO shots (id, episodeId, projectId, idx, description, dialogue, camera, assetNames, imagePrompt, videoPrompt, imageStatus, videoStatus, createdAt)
       VALUES (?,?,?,?,?,?,?,?,?,?,'none','none',?)`,
    );
    out.shots.forEach((sh, i) => {
      ins.run(
        newId(),
        episode.id,
        job.projectId,
        i + 1,
        sh.description,
        sh.dialogue,
        sh.camera,
        JSON.stringify(sh.assetNames),
        sh.imagePrompt,
        sh.videoPrompt,
        nowIso(),
      );
    });
  });
  replace();
  return `生成 ${out.shots.length} 个镜头`;
}

/** 素材图生成 */
export async function runAssetImage(job: JobRow): Promise<string> {
  const asset = db.prepare("SELECT * FROM assets WHERE id=?").get(job.targetId) as
    | { id: string; name: string; imagePrompt: string; description: string }
    | undefined;
  if (!asset) throw new Error("资产不存在");
  const prompt = asset.imagePrompt.trim() || asset.description.trim();
  if (!prompt) throw new Error("该资产没有图片提示词，请先填写");

  db.prepare("UPDATE assets SET status='running', error=NULL WHERE id=?").run(asset.id);
  try {
    const rel = await generateImage(prompt, job.projectId);
    db.prepare("UPDATE assets SET status='done', imagePath=?, error=NULL WHERE id=?").run(rel, asset.id);
    return rel;
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    db.prepare("UPDATE assets SET status='failed', error=? WHERE id=?").run(msg, asset.id);
    throw e;
  }
}

/** 镜头静帧图生成 */
export async function runShotImage(job: JobRow): Promise<string> {
  const shot = db.prepare("SELECT * FROM shots WHERE id=?").get(job.targetId) as
    | { id: string; imagePrompt: string }
    | undefined;
  if (!shot) throw new Error("镜头不存在");
  if (!shot.imagePrompt.trim()) throw new Error("该镜头没有图片提示词");

  db.prepare("UPDATE shots SET imageStatus='running', imageError=NULL WHERE id=?").run(shot.id);
  try {
    const rel = await generateImage(shot.imagePrompt, job.projectId);
    db.prepare("UPDATE shots SET imageStatus='done', imagePath=?, imageError=NULL WHERE id=?").run(rel, shot.id);
    return rel;
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    db.prepare("UPDATE shots SET imageStatus='failed', imageError=? WHERE id=?").run(msg, shot.id);
    throw e;
  }
}

/** 镜头视频生成（volcengine seedance，通道已接好、未实测） */
export async function runShotVideo(job: JobRow): Promise<string> {
  const shot = db.prepare("SELECT * FROM shots WHERE id=?").get(job.targetId) as
    | { id: string; imagePath: string | null; videoPrompt: string; description: string }
    | undefined;
  if (!shot) throw new Error("镜头不存在");
  if (!shot.imagePath) throw new Error("请先生成镜头图（视频需要首帧）");
  const prompt = shot.videoPrompt.trim() || shot.description.trim();
  if (!prompt) throw new Error("该镜头没有视频提示词");

  db.prepare("UPDATE shots SET videoStatus='running', videoError=NULL WHERE id=?").run(shot.id);
  try {
    const abs = path.join(MEDIA_DIR, shot.imagePath);
    const rel = await generateVideo(prompt, abs, job.projectId);
    db.prepare("UPDATE shots SET videoStatus='done', videoPath=?, videoError=NULL WHERE id=?").run(rel, shot.id);
    return rel;
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    db.prepare("UPDATE shots SET videoStatus='failed', videoError=? WHERE id=?").run(msg, shot.id);
    throw e;
  }
}

File: app/lib/src/engine/pipeline/runners.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import 'package:sqlite3/sqlite3.dart';
import '../db.dart';
import '../media.dart';
import '../util.dart';
import '../providers/gateway.dart';
import 'prompts.dart' as p;
import 'schemas.dart';

class JobRow {
  final String id;
  final String projectId;
  final String kind;
  final String targetId;
  final String payload;
  const JobRow(this.id, this.projectId, this.kind, this.targetId, this.payload);
}

/// 流水线执行器（移植 server/src/pipeline/runners.ts）。
/// 保留 v0.1 修复语义：媒体类执行器先置实体 running，预检在 try 内——
/// 任何失败实体必落 failed+原因，绝不卡 queued。
class Runners {
  final Database db;
  final ProviderGateway gateway;
  final MediaStore media;
  Runners(this.db, this.gateway, this.media);

  Future<String> run(JobRow job, CancelToken token) => switch (job.kind) {
        'script_gen' => _runScriptGen(job, token),
        'asset_extract' => _runAssetExtract(job, token),
        'storyboard_gen' => _runStoryboardGen(job, token),
        'asset_image' => _runAssetImage(job, token),
        'shot_image' => _runShotImage(job, token),
        'shot_video' => _runShotVideo(job, token),
        _ => throw EngineException('未知任务类型 ${job.kind}'),
      };

  // ---------- 结构化文本调用（解析失败带错误反馈自愈重试一次） ----------

  Future<T> _structuredText<T>(String system, String user,
      T Function(dynamic) parse, CancelToken token) async {
    final first = await gateway.generateText(system, user, cancelToken: token);
    try {
      return parse(extractJson(first.content));
    } catch (e1) {
      final err1 = errMessage(e1);
      final second = await gateway.generateText(
          system,
          p.repairUser(user, first.content,
              err1.length > 300 ? err1.substring(0, 300) : err1),
          cancelToken: token);
      try {
        return parse(extractJson(second.content));
      } catch (e2) {
        final err2 = errMessage(e2);
        final head = second.content.length > 200
            ? second.content.substring(0, 200)
            : second.content;
        throw EngineException(
            '模型输出两次都无法解析。最后错误: ${err2.length > 300 ? err2.substring(0, 300) : err2}。原始输出开头: $head');
      }
    }
  }

  Row _mustGetProject(String projectId) {
    final rows =
        db.select('SELECT * FROM projects WHERE id=?', [projectId]);
    if (rows.isEmpty) throw EngineException('项目不存在（可能已被删除）');
    return rows.first;
  }

  // ---------- 小说 → 分集剧本 ----------

  Future<String> _runScriptGen(JobRow job, CancelToken token) async {
    final project = _mustGetProject(job.projectId);
    final payload = jsonDecode(job.payload) as Map<String, dynamic>;
    final novels = db.select(
        'SELECT title, content FROM novels WHERE projectId=?',
        [job.projectId]);
    if (novels.isEmpty ||
        (novels.first['content'] as String).trim().isEmpty) {
      throw EngineException('请先导入小说');
    }
    final novel = novels.first;
    final episodeCount =
        ((payload['episodeCount'] as num?)?.toInt() ?? 3).clamp(1, 12);

    final out = await _structuredText(
        p.scriptGenSystem,
        p.scriptGenUser(
            (novel['title'] as String).isNotEmpty
                ? novel['title'] as String
                : project['name'] as String,
            novel['content'] as String,
            episodeCount),
        parseScriptOut,
        token);

    db.execute('BEGIN');
    try {
      db.execute('DELETE FROM episodes WHERE projectId=?', [job.projectId]);
      final ins = db.prepare(
          'INSERT INTO episodes (id, projectId, idx, title, synopsis, scriptJson, createdAt) VALUES (?,?,?,?,?,?,?)');
      for (final (i, ep) in out.episodes.indexed) {
        ins.execute([
          newId(),
          job.projectId,
          i + 1,
          ep.title,
          ep.synopsis,
          jsonEncode(ep.scenes),
          nowIso()
        ]);
      }
      ins.dispose();
      db.execute('COMMIT');
    } catch (_) {
      db.execute('ROLLBACK');
      rethrow;
    }
    return '生成 ${out.episodes.length} 集剧本';
  }

  // ---------- 全部剧本 → 资产列表 ----------

  Future<String> _runAssetExtract(JobRow job, CancelToken token) async {
    final project = _mustGetProject(job.projectId);
    final episodes = db.select(
        'SELECT title, synopsis, scriptJson FROM episodes WHERE projectId=? ORDER BY idx',
        [job.projectId]);
    if (episodes.isEmpty) throw EngineException('请先生成剧本');

    final summary = episodes.indexed
        .map((e) =>
            '第${e.$1 + 1}集《${e.$2['title']}》：${e.$2['synopsis']}\n${e.$2['scriptJson']}')
        .join('\n\n');
    final assets = await _structuredText(
        p.assetExtractSystem,
        p.assetExtractUser(summary, project['artStyle'] as String),
        parseAssetsOut,
        token);

    var added = 0, updated = 0;
    db.execute('BEGIN');
    try {
      for (final a in assets) {
        final existing = db.select(
            'SELECT id FROM assets WHERE projectId=? AND name=? AND kind=?',
            [job.projectId, a.name, a.kind]);
        if (existing.isNotEmpty) {
          db.execute(
              'UPDATE assets SET description=?, imagePrompt=? WHERE id=?',
              [a.description, a.imagePrompt, existing.first['id']]);
          updated++;
        } else {
          db.execute(
              "INSERT INTO assets (id, projectId, kind, name, description, imagePrompt, status, createdAt) VALUES (?,?,?,?,?,?,'draft',?)",
              [
                newId(),
                job.projectId,
                a.kind,
                a.name,
                a.description,
                a.imagePrompt,
                nowIso()
              ]);
          added++;
        }
      }
      db.execute('COMMIT');
    } catch (_) {
      db.execute('ROLLBACK');
      rethrow;
    }
    return '提取资产：新增 $added 个，更新 $updated 个';
  }

  // ---------- 单集剧本 → 分镜 ----------

  Future<String> _runStoryboardGen(JobRow job, CancelToken token) async {
    final project = _mustGetProject(job.projectId);
    final episodes =
        db.select('SELECT * FROM episodes WHERE id=?', [job.targetId]);
    if (episodes.isEmpty) throw EngineException('剧集不存在');
    final episode = episodes.first;

    final assets = db.select(
        'SELECT kind, name, description FROM assets WHERE projectId=?',
        [job.projectId]);
    final kindLabel = {'character': '角色', 'scene': '场景', 'prop': '道具'};
    final assetsContext = assets.isNotEmpty
        ? assets
            .map((a) =>
                '[${kindLabel[a['kind']] ?? a['kind']}] ${a['name']}：${a['description']}')
            .join('\n')
        : '（尚未提取资产，请根据剧本自行保持角色外观一致）';

    final shots = await _structuredText(
        p.storyboardGenSystem,
        p.storyboardGenUser(episode['title'] as String,
            episode['scriptJson'] as String, assetsContext,
            project['artStyle'] as String),
        parseShotsOut,
        token);

    db.execute('BEGIN');
    try {
      db.execute('DELETE FROM shots WHERE episodeId=?', [episode['id']]);
      final ins = db.prepare(
          "INSERT INTO shots (id, episodeId, projectId, idx, description, dialogue, camera, assetNames, imagePrompt, videoPrompt, imageStatus, videoStatus, createdAt) VALUES (?,?,?,?,?,?,?,?,?,?,'none','none',?)");
      for (final (i, sh) in shots.indexed) {
        ins.execute([
          newId(),
          episode['id'],
          job.projectId,
          i + 1,
          sh.description,
          sh.dialogue,
          sh.camera,
          jsonEncode(sh.assetNames),
          sh.imagePrompt,
          sh.videoPrompt,
          nowIso()
        ]);
      }
      ins.dispose();
      db.execute('COMMIT');
    } catch (_) {
      db.execute('ROLLBACK');
      rethrow;
    }
    return '生成 ${shots.length} 个镜头';
  }

  // ---------- 素材图 ----------

  Future<String> _runAssetImage(JobRow job, CancelToken token) async {
    final rows = db.select('SELECT * FROM assets WHERE id=?', [job.targetId]);
    if (rows.isEmpty) throw EngineException('资产不存在');
    final asset = rows.first;

    // 先置 running 再预检：任何失败（含预检）实体必落 failed+原因，不能卡 queued
    db.execute("UPDATE assets SET status='running', error=NULL WHERE id=?",
        [asset['id']]);
    try {
      final prompt = (asset['imagePrompt'] as String).trim().isNotEmpty
          ? (asset['imagePrompt'] as String).trim()
          : (asset['description'] as String).trim();
      if (prompt.isEmpty) throw EngineException('该资产没有图片提示词，请先填写');
      final rel = await gateway.generateImage(prompt, job.projectId,
          cancelToken: token);
      db.execute(
          "UPDATE assets SET status='done', imagePath=?, error=NULL WHERE id=?",
          [rel, asset['id']]);
      return rel;
    } catch (e) {
      final msg = errMessage(e);
      db.execute("UPDATE assets SET status='failed', error=? WHERE id=?",
          [msg.length > 2000 ? msg.substring(0, 2000) : msg, asset['id']]);
      rethrow;
    }
  }

  // ---------- 镜头静帧图 ----------

  Future<String> _runShotImage(JobRow job, CancelToken token) async {
    final rows = db.select('SELECT * FROM shots WHERE id=?', [job.targetId]);
    if (rows.isEmpty) throw EngineException('镜头不存在');
    final shot = rows.first;

    db.execute(
        "UPDATE shots SET imageStatus='running', imageError=NULL WHERE id=?",
        [shot['id']]);
    try {
      final prompt = (shot['imagePrompt'] as String).trim();
      if (prompt.isEmpty) throw EngineException('该镜头没有图片提示词');
      final rel = await gateway.generateImage(prompt, job.projectId,
          cancelToken: token);
      db.execute(
          "UPDATE shots SET imageStatus='done', imagePath=?, imageError=NULL WHERE id=?",
          [rel, shot['id']]);
      return rel;
    } catch (e) {
      final msg = errMessage(e);
      db.execute(
          "UPDATE shots SET imageStatus='failed', imageError=? WHERE id=?",
          [msg.length > 2000 ? msg.substring(0, 2000) : msg, shot['id']]);
      rethrow;
    }
  }

  // ---------- 镜头视频 ----------

  Future<String> _runShotVideo(JobRow job, CancelToken token) async {
    final rows = db.select('SELECT * FROM shots WHERE id=?', [job.targetId]);
    if (rows.isEmpty) throw EngineException('镜头不存在');
    final shot = rows.first;

    db.execute(
        "UPDATE shots SET videoStatus='running', videoError=NULL WHERE id=?",
        [shot['id']]);
    try {
      final imagePath = shot['imagePath'] as String?;
      if (imagePath == null || imagePath.isEmpty) {
        throw EngineException('请先生成镜头图（视频需要首帧）');
      }
      final prompt = (shot['videoPrompt'] as String).trim().isNotEmpty
          ? (shot['videoPrompt'] as String).trim()
          : (shot['description'] as String).trim();
      if (prompt.isEmpty) throw EngineException('该镜头没有视频提示词');
      final rel = await gateway.generateVideo(
          prompt, media.absPath(imagePath), job.projectId,
          cancelToken: token);
      db.execute(
          "UPDATE shots SET videoStatus='done', videoPath=?, videoError=NULL WHERE id=?",
          [rel, shot['id']]);
      return rel;
    } catch (e) {
      final msg = errMessage(e);
      db.execute(
          "UPDATE shots SET videoStatus='failed', videoError=? WHERE id=?",
          [msg.length > 2000 ? msg.substring(0, 2000) : msg, shot['id']]);
      rethrow;
    }
  }
}
